### ISSUE TITLE
Missing variant for Adding Entries to a Property List

### ISSUE THREAD

**Issue by car222222** [latex3/latex3#926]
It seems that we are missing a command to add an entry only in the case the key does not already exist in the list.

**Comment by car222222** [comment]
Got that one completely the wrong way around!   Now edited to the correct version.

**Comment by u-fischer** [comment]
do you have any use case where you need such a command?

**Comment by blefloch** [comment]
@u-fischer: for instance one could keep track of options (or "attributes" of an object) as follows:
```
\prop_new:N \l_@@_options_prop
\prop_set_from_keyval:Nn \l_@@_options_prop
  { width = , height = , depth = }
\cs_new_protected:Npn \@@_update_option:nn #1#2
  {
    \prop_put_if_exist:NnnF \l_@@_options_prop {#1} {#2}
      { \msg_error:nnnn { @@ } { inexistent-option } {#1} {#2} }
  }
```
I'm wondering if there could be a more general `\prop_update:Nnnn` (prop) (key) (code) (fallback) where the (code) would take the previous value associated to the (key) and "produce" the new value to be stored, while the (fallback) would be used when the (key) is not there.  Ideally we could have something general enough that extends `prop_get`, `prop_put`, `prop_put_if_new`, `prop_pop` etc.  But I don't see a precise syntax for it.

**Comment by u-fischer** [comment]
@blefloch  I know that there are cases where you want to restrict the attributes, but I wonder how often you need a check at this level. I mean I know that there are only three options and don't need a guard, and users normally  don't fill properties directly but use a key-val interface which would catch this.

**Comment by car222222** [comment]
@u-fischer I see PLs as having far more general uses, not necessarily related to keyval lists.  

On the use-case:
Maybe there is nothing direct for the crude (raw:-) version, but the following particular case of @blefloch 's suggestion would seem to be the Bees Knees for the ubiquitously important use-case of data verification:

`\prop_put_if_exist:NnnF`

Reiteration: This is unrelated to key-gals!

**Comment by car222222** [comment]
@blefloch Yes, I had also been thinking of some such generalised command to cover all cases!  
Definitely worth pursuing due to the huge potential importance of PLs.

But I suspect that some particular cases will need their own commands with simple names: thus, for example, the simple base name`\prop_update:N...` would be good to,use for the case where 
the if-key-exists code is 'change the value' and the if-key-new code is 'add new entry'.

**Comment by blefloch** [comment]
A possible syntax would be
```
\prop_update:Nnnn \l_tmpa_prop { key }
  {
    \some_code:n {#1} % non-expandable is ok
    \int_case:nnF {#1}
      {
        { 1 } { \prg_return:n { 2 } } % new value
        { 2 } { \prg_return:n { 3 } } % new value
        { 3 } { \prg_return:n { 4 } } % new value
        { 4 } { \prg_return_true: } % keep same value
      }
      { \prg_return_false: } % delete value
  }
  { \prg_return:n { 1 } } % value in case the key wasn't there
```
so that for instance `\prop_del:Nn` would be
```
\prop_update:Nnnn #1 {#2} { \prg_return_false: } { \prg_return_false: }
```
while `\prop_put_if_exist:Nnn` would be
```
\prop_update:Nnnn #1 {#2} { \prg_return:n {#3} } { \prg_return_false: }
```
